src/subtype.c

```diff
@@ -4499,7 +4499,8 @@ static jl_value_t *widen_diagonal(jl_value_t *t, jl_unionall_t *u, jl_varbinding
     if (vb.innervars != NULL) {
         for (size_t i = 0; i < jl_array_nrows(vb.innervars); i++) {
             jl_tvar_t *var = (jl_tvar_t*)jl_array_ptr_ref(vb.innervars, i);
-            nt = jl_type_unionall(var, nt);
+            if (jl_has_typevar(nt, var))
+                nt = jl_type_unionall(var, nt);
```

Comment:
I think this check is already inside `jl_type_unionall`, so is unnecessary to duplicate here